File: 프로그래머스/행렬 테두리 회전하기.py
def solution(rows, columns, queries):
    matrix = [[0 for _ in range(columns)] for _ in range(rows)]
    answer = []
    # 행렬에 1씩 증가하는 수 채워넣음
    for i in range(rows):
        for j in range(columns):
            matrix[i][j] = i * columns + j + 1
            
    for x1, y1, x2, y2 in queries:
        # 시작점은 첫번째 for문에서 값이 사라지므로 보관
        start = matrix[x1-1][y1-1]
        # 시작점은 옮기는 수들의 min 연산에 포함되지 않으므로 포함위해 초기화
        minimum = start
        # 왼쪽 변
        for i in range(x1, x2):
            number = matrix[i][y1-1]
            matrix[i-1][y1-1] = number
            minimum = min(minimum, number)
        # 아래쪽 변
        for j in range(y1, y2):
            number = matrix[x2-1][j]
            matrix[x2-1][j-1] = number
            minimum = min(minimum, number)
        # 오른쪽 변    
        for i in range(x2, x1, -1):
            number = matrix[i-2][y2-1]
            matrix[i-1][y2-1] = number
            minimum = min(minimum, number)
        # 위쪽 변
        for j in range(y2, y1, -1):
            number = matrix[x1-1][j-2]
            matrix[x1-1][j-1] = number
            minimum = min(minimum, number)
        # 시작점은 아직 옮기지 못했으므로 여기서 반영     
        matrix[x1-1][y1] = start
        answer.append(minimum)
    
    return answer

File: 프로그래머스/행렬 테두리 회전하기.java
class Solution {
    public int[] solution(int rows, int columns, int[][] queries) {
        // 정답 배열 생성
        int[] answer = new int[queries.length];
        // 초기 행렬 원소 채워서 만들기
        int[][] matrix = new int[rows][columns];
        for (int i = 0; i < rows; i++) {
            int tmp = i * columns;
            for (int j = 0; j < columns; j++)
                matrix[i][j] = tmp + j + 1;
        }
        // 행렬 테두리 회전하기
        for (int j = 0; j < queries.length; j++) {
            int[] q = queries[j];
            // 왼쪽 위 기준 숫자 보관 변수, 최소값 변수
            int keep = matrix[q[0]-1][q[1]-1], min = keep;
            // 왼쪽 변
            for (int i = q[0]; i < q[2]; i++) {
                int element = matrix[i][q[1]-1];
                matrix[i-1][q[1]-1] = element;
                min = Math.min(min, element);
            }
            // 아래쪽 변
            for (int i = q[1]; i < q[3]; i++) {
                int element = matrix[q[2]-1][i];
                matrix[q[2]-1][i-1] = element;
                min = Math.min(min, element);
            }
            // 오른쪽 변
            for (int i = q[2]-2; i >= q[0]-1; i--) {
                int element = matrix[i][q[3]-1];
                matrix[i+1][q[3]-1] = element;
                min = Math.min(min, element);
            }
            // 위쪽 변
            for (int i = q[3]-2; i >= q[1]-1; i--) {
                int element = matrix[q[0]-1][i];
                matrix[q[0]-1][i+1] = element;
                min = Math.min(min, element);
            }
            // 보관한 원소 해당 위치 삽입
            matrix[q[0]-1][q[1]] = keep;
            // 최소값 정답 배열에 넣기
            answer[j] = min;
        }
        return answer;
    }
}